src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/OffenceSdsPlusLookupService.kt

```diff
@@ -0,0 +1,141 @@
+package uk.gov.justice.digital.hmpps.calculatereleasedatesapi.service
+
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.OffenderOffence
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.SentenceAndOffences
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.SentenceCalculationType
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.util.isAfterOrEqualTo
+import java.time.LocalDate
+import java.time.Period
+import java.util.*
+
+@Service
+class OffenceSdsPlusLookupService(
+  private val manageOffencesService: ManageOffencesService,
+) {
+  val postPcscCalcTypes: Map<String, EnumSet<SentenceCalculationType>> = mapOf(
+    "SDS" to EnumSet.of(
+      SentenceCalculationType.ADIMP,
+      SentenceCalculationType.ADIMP_ORA,
+    ),
+    "DYOI" to EnumSet.of(
+      SentenceCalculationType.YOI,
+      SentenceCalculationType.YOI_ORA,
+    ),
+    "S250" to EnumSet.of(
+      SentenceCalculationType.SEC250,
+      SentenceCalculationType.SEC250_ORA,
+    ),
+  )
+
+  fun setSdsPlusMarkerForOffences(sentencesAndOffences: List<SentenceAndOffences>) {
+    val bookingIdToSentences = getMatchingSentencesToBookingId(sentencesAndOffences)
+    val offencesToCheck = getOffenceCodesToCheckWithMO(bookingIdToSentences)
+
+    if (offencesToCheck.isNotEmpty()) {
+      val moCheckResponses = manageOffencesService.getPcscMarkersForOffenceCodes(*offencesToCheck.toTypedArray()).associateBy { it.offenceCode }
+
+      bookingIdToSentences.forEach {
+        it.value.forEach { sentenceAndOffence ->
+          sentenceAndOffence.offences.filter { offenderOffence -> offencesToCheck.contains(offenderOffence.offenceCode) }
+            .forEach { offence ->
+              val moResponseForOffence = moCheckResponses[offence.offenceCode]
+              val sentenceIsAfterPcsc = sentencedAfterPcsc(sentenceAndOffence)
+              var sdsPlusIdentified = false
+              val sentenceCalculationType = SentenceCalculationType.from(sentenceAndOffence.sentenceCalculationType)
+              val sevenYearsOrMore = sevenYearsOrMore(sentenceAndOffence)
+
+              if (postPcscCalcTypes["SDS"]!!.contains(sentenceCalculationType) ||
+                postPcscCalcTypes["DYOI"]!!.contains(sentenceCalculationType)
+              ) {
+                if (sentencedWithinOriginalSdsPlusWindow(sentenceAndOffence) && sevenYearsOrMore && moResponseForOffence?.pcscMarkers?.inListA == true) {
```

Comment:
By the time I'm here, I might be a bit lost in the foreach/filter/foreach/nested if blocks...Can we pull some of these out to well named methods to make it easier to read?